File: software/examples/delphi/ExampleScanForTags.pas
program ExampleCallback;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletNFCRFID;

type
  TExample = class
  private
    ipcon: TIPConnection;
    nfc: TBrickletNFCRFID;
    tagType: byte;
  public
    procedure StateChangedCB(sender: TBrickletNFCRFID; const state: byte; const idle: boolean);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'hjw'; { Change to your UID }

var
  e: TExample;

{ Callback function for state changed callback }
procedure TExample.StateChangedCB(sender: TBrickletNFCRFID; const state: byte; const idle: boolean);
  var newTagType: byte;
  var tidLength: byte;
  var tid: Array[0..6] of byte;
  var s: String;
  var i: byte;
begin
  if idle then begin
    tagType := (tagType + 1) Mod 3;
    sender.RequestTagID(tagType);
  end;

  if state = BRICKLET_NFC_RFID_STATE_REQUEST_TAG_ID_READY then begin
    sender.GetTagID(newTagType, tidLength, tid);
	s := 'Fount tag of type ' + IntToStr(newTagType);
	s := s + ' with ID [' + IntToHex(tid[0], 2);
    if tidLength = 7 then begin
      for i:= 1 to 6 do begin
        s := s + ', ' + IntToHex(tid[i], 2);
      end;
    end
    else begin
      for i:= 1 to 3 do begin
        s := s + ', ' + IntToHex(tid[i], 2);
      end;
    end;
    s := s + ']';
    WriteLn(s);
  end;
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  nfc := TBrickletNFCRFID.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Register state changed callback to procedure StateChangedCB }
  nfc.OnStateChanged := {$ifdef FPC}@{$endif}StateChangedCB;
  
  tagType := BRICKLET_NFC_RFID_TAG_TYPE_MIFARE_CLASSIC;
  nfc.RequestTagID(tagType);

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/vbnet/ExampleScanForTags.vb
Imports System
Imports Tinkerforge

Module ExampleScanForTags
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change XYZ to the UID of your NFC/RFID Bricklet

    Private currentTagType As Byte = 0

    ' Callback subroutine for state changed callback
    Sub StateChangedCB(ByVal sender As BrickletNFCRFID, ByVal state As Byte, _
                       ByVal idle As Boolean)
        If state = BrickletNFCRFID.STATE_REQUEST_TAG_ID_READY Then
            Dim tagType As Byte
            Dim tidLength As Byte
            Dim tid() As Byte

            sender.GetTagID(tagType, tidLength, tid)

            If tidLength = 7 Then
                Console.WriteLine("Fount tag of type {0} with ID [{1:X} {2:X} {3:X} {4:X} {5:X} {6:X} {7:X}]",
                                  tagType, tid(0), tid(1), tid(2), tid(3), tid(4), tid(5), tid(6))
            Else
                Console.WriteLine("Fount tag of type {0} with ID [{1:X} {2:X} {3:X} {4:X}]",
                                  tagType, tid(0), tid(1), tid(2), tid(3))
            End If
        End If

        ' Cycle through all types
        If idle Then
            currentTagType = Convert.ToByte((currentTagType + 1) Mod 3)
            sender.requestTagID(currentTagType)
        End If
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim nr As New BrickletNFCRFID(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Register state changed callback to subroutine StateChangedCB
        AddHandler nr.StateChangedCallback, AddressOf StateChangedCB

        ' Start scan loop
        nr.RequestTagID(BrickletNFCRFID.TAG_TYPE_MIFARE_CLASSIC)

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module
